common/objectcachev2.h

```diff
@@ -0,0 +1,257 @@
+improvements:
+
+1. Less lock in both read and write.
+2. Object destruction always goes in background photon thread called reclaimer.
+3. Self adjustable timeout for reclaimer. No needs to set cycler timer.
+4. New `update` API, able to immediately substitute objects.
+5. `ObjectCacheV2` no longer support acquire/release API.
+
+It should work as `ObjectCache` in code do not depends on acquire/release API.
+
+**/
+
+template <typename K, typename VPtr>
+class ObjectCacheV2 {
+protected:
+    using V = std::remove_pointer_t<VPtr>;
+
+    struct Box : public intrusive_list_node<Box> {
+        const K key;
+        std::shared_ptr<V> ref;
+        photon::spinlock boxlock;
+        uint64_t lastcreate = 0;
+        uint64_t timestamp = 0;
+
+        struct Updater {
+            Box* box = nullptr;
+
+            ~Updater() {}
+
+            // Returned RefPtr is the old one
+            // other reader will get new one after updated
+            std::shared_ptr<V> update(V* val, uint64_t ts = 0,
+                                      std::shared_ptr<V>* newptr = nullptr) {
+                auto r = std::shared_ptr<V>(val);
+                if (newptr)
+                    std::atomic_store(newptr, r);
+                *newptr = r;
+                box->lastcreate = ts;
+                r = std::atomic_exchange(&box->ref, r);
+                return r;
+            }
+
+            explicit Updater(Box* b) : box(b) {}
+
+            Updater(const Updater&) = delete;
+            Updater(Updater&& rhs) : box(nullptr) { *this = std::move(rhs); }
+            Updater& operator=(const Updater&) = delete;
+            Updater& operator=(Updater&& rhs) {
+                std::swap(box, rhs.box);
+                return *this;
+            }
+            operator bool() const { return box != nullptr; }
+        };
+
+        Box() = default;
+        template <typename KeyType>
+        explicit Box(KeyType&& key)
+            : key(std::forward<KeyType>(key)), ref(nullptr) {}
+
+        ~Box() {}
+
+        Updater writer() { return Updater(this); }
+        std::shared_ptr<V> reader() {
+            return std::atomic_load(&ref);
+        }
+    };
+    struct ItemHash {
+        size_t operator()(const Box* x) const { return std::hash<K>()(x->key); }
+    };
+    struct ItemEqual {
+        bool operator()(const Box* a, const Box* b) const {
+            return a->key == b->key;
+        }
+    };
+
+    photon::thread* reclaimer = nullptr;
+    photon::common::RingChannel<
+        LockfreeMPMCRingQueue<std::shared_ptr<V>*, 4096>>
+        reclaim_queue;
+    photon::spinlock maplock;
+    std::unordered_set<Box*, ItemHash, ItemEqual> map;
+    intrusive_list<Box> cycle_list;
+    uint64_t lifespan;
+    photon::Timer _timer;
+
+    std::shared_ptr<V> __update(Box* item, V* val) {
+        std::shared_ptr<V> ret;
+        reclaim_queue.template send<PhotonPause>(new std::shared_ptr<V>(
```

Comment:
split it into two statements.

Reply:
```new std::shared_ptr<V>(...)``` looks evil. Try to avoid it.

Reply:
Lockfree queue accepts only POD types, a `shared_ptr<T>` object cannot sen via lockfree queue.

Maybe I can try to specialize a lockfree queue with `shared_ptr<T>` type.

Reply:
Done